File: luck_money.py
# -*- coding: utf-8 -*-
from util.wx_algorithms import *
from dicttoxml import dicttoxml
import requests
from dao.server_config import ServerConfigDao
from dao.verify_refund import VerifyRefundDao
from dao.shop_setting import ShopSettingDao
import xml.etree.ElementTree as ET
from md5 import md5

def sign(params,pay_key):
    '''
    https://pay.weixin.qq.com/wiki/doc/api/tools/cash_coupon.php?chapter=4_3
    '''
    params = [(str(key), str(val)) for key, val in params.iteritems() if val]
    sorted_params_string = '&'.join('='.join(pair) for pair in sorted(params))
    sign = '{}&key={}'.format(sorted_params_string, pay_key)
    return md5(sign).hexdigest().upper()

def sendLuckyMoney(open_id,order_id,amount,mch_id,appid,send_name,pay_key):
    nonce_str = id_generator()
    mch_billno = order_id
    mch_id = mch_id
    wxappid = appid
    send_name = send_name
    re_openid = open_id
    total_amount = amount
    total_num = '1'
    wishing = u'感谢您的好评'
    client_ip = '120.25.195.197'
    act_name = u'红包返现'
    remark = '1'

    data = {'nonce_str':nonce_str,'mch_billno':mch_billno,'mch_id':mch_id,'wxappid':wxappid,
            'send_name':send_name,'re_openid':re_openid,'total_amount':total_amount,
            'total_num':total_num,'wishing':wishing,'client_ip':client_ip,'act_name':act_name,
            'remark':remark}

    #data = {'nonce_str':nonce_str,'mch_billno':mch_billno,'mch_id':mch_id,'wxappid':wxappid,
    #        're_openid':re_openid,'total_amount':total_amount,
    #        'total_num':total_num,'client_ip':client_ip,
    #        'remark':remark}

    stringA = ''
    for key in sorted(data):
        stringA += key + '=' + data[key] + '&'
    stringA = stringA[:-1]
    stringSignTemp = stringA + "&key="+pay_key

    print stringSignTemp
    h = hashlib.md5(stringSignTemp.encode('utf-8'))
    sign = h.hexdigest().upper()
    print sign
    # sign_str = sign(data,pay_key)
    #data['wishing'] = wishing
    #data['act_name'] = act_name
    #data['send_name'] =send_name
    data['sign'] = sign
    xml = dicttoxml(data, custom_root='xml', attr_type=False)
    print xml
    url = r'https://api.mch.weixin.qq.com/mmpaymkttransfers/sendredpack'
    s = requests.post(url=url,data=xml,cert=("../pems/apiclient_cert.pem","../pems/apiclient_key.pem"))
    xml = ET.fromstring(s.text.encode('utf-8'))
    code = xml.find('result_code').text
    if code == 'SUCCESS':
        print s.text
        return True
    else:
        print s.text
        return False


if __name__ == '__main__':
    payKey = ServerConfigDao()['pay_key']
    payId = ServerConfigDao()['pay_id']
    appid = ServerConfigDao()['app_id']
    shopIdtoName = {}

    while True:
        rows = VerifyRefundDao().selectByPage(100)
        if len(rows) == 0:
            break
        for row in rows:
            if row['shop_id'] not in shopIdtoName:
                shopSetting = ShopSettingDao().getSetting(row['shop_id'])
                shopIdtoName[row['shop_id']] = shopSetting['name']
            shop_name = shopIdtoName[row['shop_id']]
            order_id = row['order_id']
            open_id = row['open_id']
            result = sendLuckyMoney(open_id,order_id,str(row['money']),payId,appid,shop_name,payKey)
            if result :
                VerifyRefundDao().setRefundFlag(row['id'])

File: luck_money.php
<?php
require 'wxpay.class.php';//数组参数
$mysql_server_name='localhost'; //
 
$mysql_username='xiaob'; //
 
$mysql_password='skdfjkasdf'; //
 
$mysql_database='xiaob'; //

$conn=mysql_connect($mysql_server_name,$mysql_username,$mysql_password) or die("error connecting") ; //连接数据库
mysql_query("set names 'utf8'"); //UTF-8 国际标准编码.
mysql_select_db($mysql_database); //打开数据库
$stopFlag = false;
$lastRow = 0;
while(! $stopFlag){
    $config = array();
    $sql = "select * from server_config where name_space='we.xwpay.sickworm.com'";
    $result2 = mysql_query($sql,$conn);
    while($row =  mysql_fetch_array($result2))
    {
        $config[$row['k']] = $row['v'];   
    }

    $sql = 'select * from verify_refund where refund_flag=0 and del_flag=0 order by id asc limit 100';
    $result = mysql_query($sql,$conn);
    $stopFlag = true;
    while($row = mysql_fetch_array($result))
    {
	$stopFlag = false;
	if ($row['id'] <= $lastRow){
	    $stopFlag = true;
	    break;
	}
        $sql = 'select * from shop_setting where shop_id='.$row['shop_id'];
        $result2 = mysql_query($sql,$conn);
        $ss = mysql_fetch_row($result2);
        $shop_name = $ss[1];
        $order_id = $row['order_id'];
        $open_id = $row['open_id'];
        $money = $row['money'];
        $sender = $shop_name;

	$obj2 = array();
	$obj2['wxappid'] = $config['app_id']; //appid
	$obj2['mch_id'] = $config['pay_id'];//商户id
	$obj2['mch_billno'] = $order_id;
	$obj2['client_ip'] = '120.25.195.197';
	$obj2['re_openid'] = $open_id;//接收红包openid
	$obj2['total_amount'] = $money;
	$obj2['total_num'] = 1;
	$obj2['send_name'] = $shop_name;
	$obj2['wishing'] = "感谢您的好评";
	$obj2['act_name'] = $shop_name."的好评返现";
	$obj2['remark'] = $shop_name."红包";
 
	$url = "https://api.mch.weixin.qq.com/mmpaymkttransfers/sendredpack";
	$wxpay = new wxPay();
	$res = $wxpay->pay($url, $obj2,$config['pay_key']);
	if ($res){
	    $sql = 'update verify_refund set refund_flag=1 where id='.$row['id'] ;
	    mysql_query($sql);
	}
	$lastRow = $row['id'];
	
    }
}
